## Promise 一旦决议，不可修改

## Promise.all([a,b])实际上返回的是第二次 promise(a 或者 b),不是由 Promise.all()创建的第一个 Promise

## Promise.all 的模拟

```js
function add(getX, getY, cb) {
  getX((xVal) => {
    x = xVal;
    if (y !== undefined) {
      cb(x + y);
    }
  });
  getX((yVal) => {
    y = yVal;
    if (x !== undefined) {
      cb(x + y);
    }
  });
}
add(fetchX, fetchY, (sum) => {
  console.log(sum);
});
// 很接近promise.all了

function add(promiseX, promiseY) {
  return Promise.all([promiseX, promiseY]).then((res) => {
    console.log(res[0] + res[1]);
  });
}
```

## 事件侦听来实现控制反转的反转

```js
function foo(x) {
  //构造侦听器
  return listener;
}
var evt = foo(42);
evt.on("compelete", () => {
  console.log(233);
});

// 控制权恢复
var evt = foo(42);
bar(evt);


// evt就是promise的一个模拟，考虑
function foo(x) {
  return new Promise(res,rej) {

  }
}
```

## 回调本身就是一种控制反转，对控制反转的反转，称为反控制反转

## 无法使用 p instanceof Promise 来判断是否 Promise 实例，因为浏览器自己 Promise 可能跟当前窗口 iframe 的不同；妥协只能使用 thenable；所以现在类不能拥有 then 函数，否则这个值在 Promise 系统中可能会被认为的 thenable 而出错

## 回调的五大问题

- 调用过早。promise 永远异步，解决了这个问题，不会出现回调同步调用的问题
- 调用过晚。promise 调用方式保证了这一点

```js
p.then(function () {
  p.then(function () {
    console.log("C");
  });
  console.log("A");
});
p.then(function () {
  console.log("B");
});
// c的调用不会在B之前
```

- 回调未调用。回调中如果出错，可能看不到结果，promise 不会

```js
// Promise可以使用竞态来防止未调用
function timeoutPromise(delay) {
  return new Promise((res)=>{
    setTimeout(()=>{
      reject('timeout')
    },delay)
  })
}
Promise.all([timeoutPromise(1000),.. ])
```

- 调用次数过少或过多。Promise 的定义方式保证了只调用一次
- 未能传递参数/环境值。promise 至多只能有一个决议值
- 吞掉错误或异常。

```js
var p = new Promise((res) => {
  res(42);
});
p.then(
  (res) => {
    throw new Error("741");
  },
  (err) => {
    // 这里永远不会执行。返回了另外一个promise
    console.log(err);
  }
);
```

## 如果向 Promise.resolve 传递非 Promise、非 thenable 值，就会得到一个用这个值填充的 Promise；如果传递的是一个真正的 Promise，就会返回同一个 Promise

```js
// 建议总是对thenable（但是不一定是Promise）包裹一层resolve，更安全
// bad
foo(42).then(function (v) {
  console.log(v);
});
// good
Promise.resolve(foo(42)).then(function (v) {
  console.log(v);
});
```

## 每次对 promise 调用 then 都会返回一个新的 promise

## then 不管返回值是啥，都会被链接为完成

## promise 可以使用 return

## 延迟 Promise 创建一般化

```js
function delay(time) {
  return new Promise((res,rej)=> {
    setTimeout(()=>{
      res()
    },time)
  }
}
```

## promise-await ajax [Promise 化]

```js
function request(url) {
  return new Promise((res) => {
    ajax(url, res);
  });
}

// 使用
request(url).then((data) => {});
```

## 如果 promise 第 2 步出错，第三步捕获的话，第 4 步又重回正常运作

## 如果不处理错误，则会继续给到下一步,直到遇到显式处理；如果第 2 参数不是个有效值，相当于没处理

```js
var p = new Promise((res, rej) => {
  rej("oops");
});
var p2 = p.then(() => {}, null);
var p3 = p2.then(
  () => {},
  (info) => {
    console.log(info);
  }
);
```

## 为什么 promise 第一个参数，使用 resolve(决议)，而不是 fulfill（完成）；因为 resolve 可以是拒绝

```js
var rejectedTh = {
  then: function (resolved, rejected) {
    rejected("oops");
  },
};
Promise.resolve(rejectedTh);
```

## Promise 的 reject 不会像 resolve 那样自动展开；但是如果是 thenable 则会自动展开

## error-first 回调风格的错误处理

```js
function foo(cb) {
  try {
    var x = bar().baz(); //如果 bar().baz()是异步的错误则永远 无法捕捉到
    cb(null, c);
  } catch (e) {
    cb(e);
  }
}

foo((err, data) => {
  if (err) {
  }
});
```

## 如果错误阻碍了 Promise 的构造，会立即抛出异常而不是被拒绝的 Promise

- new Promise(null)
- Promise.all()
- Promise.race(42)

## 传递给 Promise.all 可以是 Promise，thenable 或者立即值，都会被 Promise.resolve 过滤

## Promise.race 传入空数组将永远不会决议

## Promise.observe 实现

```js
if (!Promise.observe) {
  Promise.observe = function (pr, cb) {
    pr.then(
      (msg) => {
        Promise.resolve(msg).then(cb);
      },
      () => {
        Promise.resolve(msg).then(cb);
      }
    );
    return pr;
  };
}
```

## Promise.first 实现

```js
if (!Promise.first) {
  Promise.first = function (prs) {
    return new Promise((resolve) => {
      prs.forEach((pr) => {
        pr.then((data) => {
          resolve(data);
        });
      });
    });
  };
}
```

## Promise.first 尝试添加所有 promise 拒绝时，发出 reject 的功能

## Promise.map 实现

```js
// 应用
var p1 = Promise.resolve(11);
var p2 = Promise.resolve(42);
Promise.map([p1, p2], (pr, done) => {
  Promise.resolve(pr).then((v) => done(v * 2), done);
});

Promise.map = function (vals, cb) {
  return Promise.all(
    vals.map((val) => {
      cb(val);
    })
  );
};

// 实现
Promise.map = function (vals, cb) {
  return Promise.all(
    vals.map(
      (val) =>
        new Promise((resolve) => {
          cb(val, resolve);
        })
    )
  );
};
```

## Promise(...) 必须和 new 一起使用

## p.catch(rejected)等价于 p.then(null,rejected)

## Promise.all([])会立即完成；Promise.race([])会挂住

## 如果链上任何一个步骤进行了自身的错误处理，p.catch 就不会得到通知

## 一个 Promise 数组可能优于单个 Promise，更符合 Promise 的理念

## 将函数参数数组解构

```js
function spread(fn) {
  return Function.apply.bind(fn, null);
}
Promise.all(foo(11, 12)).then(spread((x, y) => {}));

// es6写法 Promise.all(foo(11, 12)).then(spread([x, y] => {}));
```

## Promise.wrap 实现

```js
Promise.wrap = function (fn) {
  return function () {
    var args = [].slice.call(arguments);
    return new Promise(function (resolve, reject) {
      fn.apply(
        null,
        args.concat(function (err, data) {
          if (err) {
            reject(err);
          } else {
            resolve(data);
          }
        })
      );
    });
  };
};
```

## Promise 无法取消，否则违背未来值的可信任性
